004-isa_restructuring.org

```diff
@@ -0,0 +1,787 @@
+changed, *neither did the original behaviour*. However, future
+smartcontracts that want to make use of a revised =Mint= are merely
+encouraged to make use of the latest versions of libraries, thus
+automatically giving the developers guidance on the semantic changes.
+
+The astute reader may now have noticed that the instructions included
+in variant 1 are all linear operations with a short-circuit behaviour
+on a failed assertion.  Variant 2 makes use of this, and deliberately
+eschews any form of control flow other than assertions.  The executor
+and the smartcontracts both are capable of fast control flow, being
+either compiled, or highly-optimised interpreted languages (see WASM
+replacement RFC).  To implement fast conditional logic would require a
+huge time investment and would be a major change to the behaviour of
+Iroha.  And such a change would only be useful if and only if Iroha
+were to have its own virtual execution environment.
+
+Then one can ask what to do with =InstructionBox= variants such as
+=Transfer= which entail multiple modifications and which exist
+primarily to allow granular permissions: one cannot mint new assets,
+but one can transfer them.  In case of non-mint-able assets, one can
+also raise the question of whether or not it's possible to guarantee a
+social contract: if a certain asset is meant to be of fixed supply,
+minting more of it is a crime.  The answer to that question is not
+simple: true, the original implementation enforced this social
+contract host-side.  This logic is still present, it is however moved
+into the executor, the compromise of which is still a bigger problem
+than non-mint-able assets being mint-able. [fn:3]
+
+As a closing remark, it would be prudent to motivate the move to a
+RISC instruction set from one more point of view. CISC as a CPU
+architecture emphasises hardware, assuming that changing the hardware
+is easier than it is to change the software that runs on it.  But to
+quote a famous [[https://cs.stanford.edu/people/eroberts/courses/soco/projects/risc/risccisc/][discussion of the two approaches]],
+
+#+BEGIN_QUOTE
+The Overall RISC Advantage Today, is that the Intel x86 is arguably
+the only chip which retains CISC architecture. This is primarily due
+to advancements in other areas of computer technology. The price of
+RAM has decreased dramatically. In 1977, 1MB of DRAM cost about
+$5,000. By 1994, the same amount of memory cost only $6 (when adjusted
+for inflation). Compiler technology has also become more
+sophisticated, so that the RISC use of RAM and emphasis on software
+has become ideal.
+#+END_QUOTE
+
+As is the case and experience with Iroha v1 deployments, emphasis on
+on-chain "software" is a road to a safer and more malleable
+blockchain.  Due to the simplicity of the RISC approach, it falls in
+the realm of programs that have "obviously no bugs", rather than
+programs that "have no obvious bugs". In a reduced instruction set,
```

Comment:
```suggestion
the realm of programs that have "obviously no bugs", rather than
programs that have "no obvious bugs". In a reduced instruction set,
```
```suggestion
the realm of programs that "have obviously no bugs", rather than
programs that "have no obvious bugs". In a reduced instruction set,
```